python/JSONtoObj.c

```diff
@@ -119,6 +119,15 @@ static JSOBJ Object_newUnsignedLong(void *prv, JSUINT64 value)
   return PyLong_FromUnsignedLongLong (value);
 }
 
+static JSOBJ Object_newIntegerFromString(void *prv, char *value, size_t length)
+{
+  // PyLong_FromString requires a NUL-terminated string in CPython, contrary to the documentation: https://github.com/python/cpython/issues/59200
+  char *buf = PyObject_Malloc(length + 1);
+  memcpy(buf, value, length);
+  buf[length] = '\0';
+  return PyLong_FromString(buf, NULL, 10);
+}
+
```

Comment:
The `memcpy` here is certainly not ideal. It would probably be faster to buffer the next byte, replace it with a NUL, feed the char array through `PyLong_FromString`, and then restore it. I think that should always be possible/safe without buffer overruns, but I haven't spent a great deal of thought about it yet. Do you think it's worth exploring that sort of hack?

Reply:
Ooh, now that's a disgusting but tantalising thing to do. Two ways I imagine that causing damage are:
1. The integer we're parsing is the last part of the JSON so the next byte would be out of bounds.
2. Someone is reading the JSON string in multiple threads and the second thread happens to look at that byte when it's temporarily nullified.

1 is moot if the string we're parsing is NULL terminated (which I'm fairly certain that they already are. 2 feels like a very nasty way to derail something.

Reply:
The relevant part of the decoder boils down to:

1. If the argument is a string, call `PyUnicode_AsUTF8String` to convert to `bytes`.
2. Call `PyBytes_AsString`, which returns a NULL-terminated `char*` that is the internal buffer of the `bytes` object.
3. Iterate over that buffer.

So no out-of-bounds issue. As for multithreading, yes, that could be an issue. But on the other hand, ujson already isn't thread-safe as far as I can tell. For example, modifying a dict from another thread while encoding it should crash the encoder.

Reply:
I ran some tests on this, parsing an array equivalent to `[2**64 + i for i in range(1000)]`, i.e. 1000 ints all using this code path. On my machine, parsing such a string averages 128 μs with `memcpy` and 108 μs with the suggested hack.

Compared to the other code paths for smaller ints (note that the number of digits obviously has a direct impact on throughput as well):

input | time
-|-
`2**64 + i`, `memcpy` | 128 μs
`2**64 + i`, buffer hack | 108 μs
`2**62 + i` (`newUnsignedLong`) | 60 μs
`-2**63 + i` (`newLong`) | 64 μs
`2**30 + i` (`newInt`) | 45 μs
`-2**31 + i` (`newInt`) | 45 μs

Command, for reference: `python3 -m timeit -s 'import ujson; s = "[" + ",".join(str(2**64 + i) for i in range(1000)) + "]"' 'ujson.loads(s)'`

Implementation of the hack:

```c
static JSOBJ Object_newIntegerFromString(void *prv, char *value, size_t length)
{
  char buf = value[length];
  value[length] = '\0';
  PyObject *out = PyLong_FromString(value, NULL, 10);
  value[length] = buf;
  return out;
}
```

Reply:
That's a pretty small speedup really. On my machine it's even less (115μs vs 110μs). Definitely not worth the hack in my mind.

Reply:
Yeah, agree. It's basically only relevant if all you ever do is parse huge integers. I don't think that's common enough to warrant the ugly hack.